### ISSUE TITLE
typecheck fails incorrectly when encountering package paths with hyphens in them

### ISSUE THREAD

**Issue by lazeratops** [golangci/golangci-lint#2232]
### Welcome

- [X] Yes, I'm using a binary release within 2 latest major releases. Only such installations are supported.
- [X] Yes, I've searched similar issues on GitHub and didn't find any.
- [X] Yes, I've included all information below (version, config, etc).
- [X] Yes, I've tried with the standalone linter if available. (https://golangci-lint.run/usage/linters/)

### Description of the problem

When a package with a hyphen in the path is imported, `typecheck` returns unexpected failures regarding:

* the package not being used (it is, by its package name)
* an undeclared name in the use of the package (the name is declared within the imported package)

Example: import `"github.com/hashicorp/go-retryablehttp"`, and use `retryablehttp.NewClient()`

### Version of golangci-lint

<details>

```console
$ golangci-lint --version
golangci-lint has version 1.42.1 built from 54f4301d on 2021-09-06T17:05:10Z
```

</details>

### Configuration file

<details>

```console
$ cat .golangci.yml
run:
  timeout: 5m
  skip-dirs:
    - third-party
linters:
  enable:
    - revive

issues:
  exclude-use-default: false
  exclude:
    # EXC0001 errcheck: Almost all programs ignore errors on these functions and in most cases it's ok,
    # excluded by default
    - Error return value of .((os\.)?std(out|err)\..*|.*Close|.*Flush|os\.Remove(All)?|.*print(f|ln)?|os\.(Un)?Setenv). is not checked
```

</details>

### Go environment

<details>

```console
$ go version && go env
go version go1.16.6 linux/amd64
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/me/.cache/go-build"
GOENV="/home/me/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/me/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/me/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/bin/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/bin/go/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.16.6"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/mnt/d/git/src/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build3787202721=/tmp/go-build -gno-record-gcc-switches"
```

</details>

### Verbose output of running

<details>

```console
$ golangci-lint cache clean
$ golangci-lint run -v
INFO [config_reader] Config search paths: [./ /mnt/d/git/src /mnt/d/git /mnt/d /mnt / /home/me]
INFO [config_reader] Used config file .golangci.yaml
INFO [lintersdb] Active 11 linters: [deadcode errcheck gosimple govet ineffassign revive staticcheck structcheck typecheck unused varcheck]
INFO [loader] Go packages loading at mode 575 (exports_file|files|name|compiled_files|deps|imports|types_sizes) took 32.0411111s
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 116.9032ms
INFO [linters context/goanalysis] analyzers took 8m37.703045s with top 10 stages: buildir: 2m48.6350008s, the_only_name: 38.5943683s, nilness: 18.5184142s, typedness: 12.9233481s, fact_purity: 12.3873141s, fact_deprecated: 11.9073477s, ctrlflow: 11.7949589s, printf
: 11.6356773s, SA5012: 10.7976342s, unused: 7.7442258s
INFO [runner/skip dirs] Skipped 164 issues from dir third-party/bazel/vendored/src/main/protobuf by pattern third-party
INFO [runner/skip dirs] Skipped 4 issues from dir third-party/bazel/importer by pattern third-party
INFO [runner/skip dirs] Skipped 2 issues from dir lib/server/go/examples/grpc/server_b by pattern (^|/)examples($|/)
INFO [runner/skip dirs] Skipped 3 issues from dir lib/server/go/examples/unix-domain-socket by pattern (^|/)examples($|/)
INFO [runner/skip dirs] Skipped 1 issues from dir lib/server/go/examples/grpc/server_a by pattern (^|/)examples($|/)
INFO [runner/skip dirs] Skipped 1 issues from dir lib/server/go/examples/kubernetes-api by pattern (^|/)examples($|/)
INFO [runner] Issues before processing: 736, after processing: 3
INFO [runner] Processors filtering stat (out/in): skip_dirs: 561/736, max_from_linter: 3/3, cgo: 736/736, exclude-rules: 468/468, nolint: 417/468, uniq_by_line: 3/417, max_same_issues: 3/3, source_code: 3/3, sort_results: 3/3, exclude: 468/558, path_prettifier: 736
/736, identifier_marker: 558/558, diff: 3/3, max_per_file_from_linter: 3/3, path_shortener: 3/3, filename_unadjuster: 736/736, autogenerated_exclude: 558/561, severity-rules: 3/3, path_prefixer: 3/3, skip_files: 736/736
INFO [runner] processing took 54.6805ms with stages: nolint: 18.6559ms, path_prettifier: 17.108ms, autogenerated_exclude: 10.1485ms, identifier_marker: 6.1073ms, skip_dirs: 1.8877ms, exclude: 472.3µs, source_code: 178µs, cgo: 50.3µs, filename_unadjuster: 30.2µs, un
iq_by_line: 26.1µs, max_same_issues: 4.5µs, path_shortener: 2.1µs, max_from_linter: 1.9µs, max_per_file_from_linter: 1.6µs, skip_files: 1.4µs, sort_results: 1.2µs, diff: 1.2µs, severity-rules: 900ns, exclude-rules: 800ns, path_prefixer: 600ns
INFO [runner] linters took 19.6510716s with stages: goanalysis_metalinter: 19.5958253s
services/buildkite-notify/pkg/slack/slack.go:40:10: undeclared name: `retryablehttp` (typecheck)
        http := retryablehttp.NewClient()
                ^
services/buildkite-notify/pkg/slack/slack.go:47:14: undeclared name: `retryablehttp` (typecheck)
        req, err := retryablehttp.NewRequest("POST", u.String(), bytes.NewReader(requestBody))
                    ^
services/buildkite-notify/pkg/slack/slack.go:11:2: "github.com/hashicorp/go-retryablehttp" imported but not used (typecheck)
        "github.com/hashicorp/go-retryablehttp"
        ^
INFO File cache stats: 1 entries of total size 1.3KiB
INFO Memory: 511 samples, avg is 632.4MB, max is 2890.5MB
INFO Execution took 52.2296148s
```

</details>

### Code example or link to a public repository

<details>

```go
import (
	"github.com/hashicorp/go-retryablehttp"
)

func main() {
  _ = retryablehttp.NewClient()
}
```

</details>

**Comment by ldez** [comment]
Hello,

`typecheck` is not real linter but it's just a way to report "compilation" errors.


I don't reproduce your problem:
```go
package main

import (
	"github.com/hashicorp/go-retryablehttp"
)

func main() {
	_ = retryablehttp.NewClient()
}
```

```go
module github.com/example/example

go 1.16

require github.com/hashicorp/go-retryablehttp v0.7.0
```

```yml
run:
  timeout: 5m
  skip-dirs:
    - third-party
linters:
  enable:
    - revive

issues:
  exclude-use-default: false
  exclude:
    # EXC0001 errcheck: Almost all programs ignore errors on these functions and in most cases it's ok,
    # excluded by default
    - Error return value of .((os\.)?std(out|err)\..*|.*Close|.*Flush|os\.Remove(All)?|.*print(f|ln)?|os\.(Un)?Setenv). is not checked
```


```console
 $ golangci-lint run -v

INFO [config_reader] Used config file .golangci.yml 
INFO [lintersdb] Active 11 linters: [deadcode errcheck gosimple govet ineffassign revive staticcheck structcheck typecheck unused varcheck] 
INFO [loader] Go packages loading at mode 575 (files|types_sizes|deps|exports_file|name|compiled_files|imports) took 131.941177ms 
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 119.9µs 
INFO [linters context/goanalysis] analyzers took 0s with no stages 
INFO [runner] processing took 3.64µs with stages: max_same_issues: 623ns, max_from_linter: 360ns, nolint: 344ns, skip_dirs: 302ns, sort_results: 247ns, exclude-rules: 194ns, cgo: 187ns, autogenerated_exclude: 144ns, path_prettifier: 140ns, uniq_by_line: 136ns, identifier_marker: 130ns, exclude: 127ns, source_code: 127ns, diff: 117ns, filename_unadjuster: 116ns, skip_files: 113ns, max_per_file_from_linter: 73ns, path_shortener: 63ns, severity-rules: 49ns, path_prefixer: 48ns 
INFO [runner] linters took 36.950444ms with stages: goanalysis_metalinter: 36.887665ms 
INFO File cache stats: 0 entries of total size 0B 
INFO Memory: 3 samples, avg is 28.3MB, max is 36.6MB 
INFO Execution took 173.6119ms  
```

**Comment by lazeratops** [comment]
Ah, thanks for taking a look. I see this issue both locally and on our CI, which is weird because everything actually builds fine. If it is not reproducible I guess it is something with our environment, I'll keep digging into it on my end.

**Comment by lazeratops** [comment]
I've attempted to just disable typecheck for the time being, but it seems like it still runs even though verbose output shows that it's no longer in active linters...I am guessing this isn't reproable on your end either? Wondering if this is part of the larger typecheck problem in our environment or something else...

```
INFO [lintersdb] Active 10 linters: [deadcode errcheck gosimple govet ineffassign revive staticcheck structcheck unused varcheck]
```

**Comment by lazeratops** [comment]
Update: go get on the specific hyphenated packages resolved our issue, not yet sure why IDE was not complaining about this.

Followup suggestion: if `typecheck` is not a real linter and cannot be disabled, suggest removing it from the list of enabled linters in verbose mode and removing it if you try to disable it from the yaml configuration. It makes it look like it should be able to be enabled/disabled like the other linters, when that doesn't seem to be the case.

Thanks!
